udata/api/cache.py

```diff
@@ -0,0 +1,118 @@
+
+    Differs from the official Flask Caching Decorator
+    http://flask.pocoo.org/docs/0.12/patterns/viewdecorators/#caching-decorator
+    by dealing with appropriated headers for server AND client-side caches.
+
+    Key points:
+    * Generates the cache key using the full path (using the query string).
+    * Uses `Cache-Control`, `ETag` and `Last-Modified` headers.
+    * Respects browser's hard-refresh.
+
+    Parameters:
+    * `class_instance`: allows the decorator to be applied on a class method.
+    * `check_serverside`: either the client verifies with the server if
+      its cached versions are still valid or not.
+    * `personal`: set to `True` if the response contains sensitive
+      informations only pertinent for a given user.
+    * `client_timeout`: time in seconds before the client considers the
+      cache as being invalid.
+    * `server_timeout`: time in seconds before the server considers the
+      cache as being invalid.
+    * `key_pattern`: you can customize the cache key, it can be either
+      a callable or a string. If it contains `%s`, the placeholder will
+      be replaced by the full path (including query string).
+    * `make_response`: a way to override the default Flask way of
+      creating a response object.
+
+    Inspirations:
+    * http://codereview.stackexchange.com/questions/147038/↩︎
+      improving-the-flask-cache-decorator
+    * https://gist.github.com/glenrobertson/954da3acec84606885f5
+
+    Resources:
+    * https://jakearchibald.com/2016/caching-best-practices/
+    * https://developers.google.com/web/fundamentals/performance/↩︎
+      optimizing-content-efficiency/http-caching
+
+    Image:
+    * https://developers.google.com/web/fundamentals/performance/↩︎
+      optimizing-content-efficiency/images/http-cache-decision-tree.png
+    """
+    def _generate_cache_control():
+        cache_control = ResponseCacheControl()
+        # Before you think the logic is reversed, read Jake's article.
+        if check_serverside:
+            cache_control.no_cache = True
+        else:
+            cache_control.must_revalidate = True
+        if personal:
+            cache_control.private = True
+        else:
+            cache_control.public = True
+        cache_control.max_age = client_timeout
+        return cache_control
+
+    def _generate_cache_key(str_or_callable):
+        if callable(str_or_callable):
+            return str_or_callable()
+        else:
+            try:
+                # Include querystring with `full_path`.
+                return str_or_callable % request.full_path
+            except TypeError:
+                return str_or_callable
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            headers = Headers()
+            headers.add('Cache-Control', _generate_cache_control())
+            cache_key = _generate_cache_key(key_pattern)
+            resp = cache.get(cache_key)
```

Comment:
Let's refocus on context here… I'm not trying to send a 🚀 to the 🌔 but to cache 2 API calls which are not user-related in order to improve **incrementally** performances for the user experience.

Cache is a poor-man performance solution but we have to stick to it because the server *currently* takes [576 ms](https://www.webpagetest.org/result/170301_TJ_TZB/1/details/cached/#waterfall_view_step1) for each and every request to a dataset view.

We can wait 6 months to release a perfect fine-tuned cache solution with Varnish and all but our users will suffer during that time. I'm just trying to reduce that pain during that interval.